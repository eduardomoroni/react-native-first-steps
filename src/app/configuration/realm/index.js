import _ from 'lodash'
import * as Schemas from './schemas'

export const schemas = _.flatMap(Schemas)
export const defaultConfig = { schema: schemas, path: 'MTG-BD.realm' }
