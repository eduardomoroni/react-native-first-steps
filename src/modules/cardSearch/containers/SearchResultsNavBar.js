// @flow

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { navBackButton, navTitle, navButtonBar, openDrawer } from '../../modules/navigation/NavBarItems'
import { switchDisplayMode as switchDisplayModeAction } from '../../redux/actions'
import Menu from './ListCardFilterMenu'
import CustomNavBar from '../../modules/navigation/CustomNavBar'
import type { Dispatch } from 'redux'

export class SearchResultsNavBar extends Component {
  render () {
    const { showCardsAs, switchDisplayMode } = this.props

    const settingsButton = {
      buttonProps: {name: 'sliders'},
      callback: () => openDrawer(<Menu />)
    }

    const displayModeButton = {
      buttonProps: {name: showCardsAs === 'image' ? 'th-list' : 'th-large'},
      callback: () => switchDisplayMode(showCardsAs === 'image' ? 'list' : 'image')
    }

    return (
      <CustomNavBar
        leftRender={navBackButton()}
        middleRender={navTitle('searching_cards')}
        rightRender={navButtonBar([displayModeButton, settingsButton])} />
    )
  }
}

const mapStateToProps = (state) => {
  const { showCardsAs } = state.cardSearch
  return {
    showCardsAs
  }
}

const mapDispatchToProps = (dispatch: Dispatch<*>) => {
  return {
    switchDisplayMode: (toggle) => dispatch(switchDisplayModeAction(toggle))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultsNavBar)
