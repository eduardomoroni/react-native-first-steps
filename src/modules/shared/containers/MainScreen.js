// @flow

import React, { Component } from 'react'
import {
  View,
  StyleSheet
} from 'react-native'
import { Colors } from '../modules/theme/themes'

export default class MainScreen extends Component {
  render () {
    return (
      <View style={styles.container} />
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: Colors.background
  }
})
