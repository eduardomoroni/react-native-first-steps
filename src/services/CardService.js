import _ from 'lodash'
import { cardType } from '../types'
import * as RealmService from '../realm/RealmService.js'
import {
  jsonToRealmCard,
  placeholdersToSymbols,
  convertCardFormToRealmQueries
} from '../realm/conversion'

const upsertCard = (card) => {
  RealmService.create('Card', card, true)
}

const findCardByID = (multiverseid: number) : cardType => {
  return RealmService.objectForPrimaryKey('Card', multiverseid)
}

const findAllCards = () => {
  return RealmService.findAll('Card')
}

const findCardIndex = (results: any, card: cardType) => {
  return RealmService.findIndex(results, (obj) => obj.id === card.id)
}

const sortCards = (cards, sorting) => {
  const { field, reversed } = sorting.sortBy
  return cards.sorted(field, reversed)
}

const findCardsFromForm = (form) => {
  const realmQueries = convertCardFormToRealmQueries(form)
  let results = RealmService.findAll('Card')

  _.each(realmQueries, (query) => {
    if (query !== undefined && query.length > 0) {
      results = results.filtered(query)
    }
  })

  return results
}

const importMTGJSON = (mtgJson) => {
  mtgJson.cards.forEach((card) => {
    card.text = placeholdersToSymbols(card.text)
    const cardAsRealmObject = jsonToRealmCard(card)
    try {
      upsertCard(cardAsRealmObject)
    } catch (e) {
      console.error(`Failed to insert ${card.name}:`, e)
    }
  })
}

export {
  findCardByID,
  sortCards,
  findCardsFromForm,
  importMTGJSON,
  upsertCard,
  findAllCards,
  findCardIndex
}
