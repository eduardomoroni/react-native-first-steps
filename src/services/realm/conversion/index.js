export * from './CardForm'
export * from './JsonCard'
export * from './Placeholder'
export * from './RealmRepresentation'
