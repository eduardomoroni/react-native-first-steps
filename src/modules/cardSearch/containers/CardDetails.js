// @flow

import React, { Component } from 'react'
import { Text, View, ScrollView } from 'react-native'
import { styles } from '../../theme/cardSearch/CardDetailsStyle'
import { CardImage } from '../../shared/components/index'
import { cardType, rulingsType } from '../../../types/CardType'
import { rulingsTextAsArray } from '../../../services/realm/conversion/RealmRepresentation'
import I18n from 'react-native-i18n'

export default class CardDetails extends Component {
  renderText (value: string, key: number) {
    return (
      <Text key={key} >{`
        - ${value}`
      }</Text>
    )
  }

  renderRulings (rulings: rulingsType) {
    const rulesText = rulingsTextAsArray(rulings)

    return (
      <View style={styles.rowContainer}>
        <Text style={styles.rowText}>
          <Text style={styles.rowName}>{`${I18n.t('rulings')}: `}</Text>
          { rulesText.map(this.renderText) }
        </Text>
      </View>
    )
  }

  renderRow (rowName: string, rowText: string) {
    if (rowText === null) {
      return null
    }

    return (
      <View style={styles.rowContainer}>
        <Text style={styles.rowText}>
          <Text style={styles.rowName}>{`${I18n.t(rowName)}: `}</Text>
          {rowText}
        </Text>
      </View>
    )
  }

  render () {
    const { card } = this.props

    return (
      <ScrollView style={styles.container}>
        <View style={styles.card}>
          <CardImage card={card} />
        </View>
        {this.renderRow('name', card.name)}
        {this.renderRow('type', card.type)}
        {this.renderRow('P/T', `${card.power}/${card.toughness}`)}
        {this.renderRow('manaCost', `${card.manaCost} (${card.cmc})`)}
        {this.renderRow('rarity', card.rarity)}
        {this.renderRow('text', card.text)}
        {this.renderRow('flavor', card.flavor)}
        {this.renderRulings(card.rulings)}
        {this.renderRow('sets', 'Estou com pregui√ßa')}
        {this.renderRow('legality', 'Farei depois')}
        {this.renderRow('foreignNames', 'Um dia')}
        <View style={styles.card} />
      </ScrollView>
    )
  }
}

CardDetails.propTypes = {
  card: cardType
}
