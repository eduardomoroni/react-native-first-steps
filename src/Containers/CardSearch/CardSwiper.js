import React, { Component } from 'react'
import PropTypes from 'prop-types'
import CardDetails from './CardDetails'
import { Actions as NavigationActions } from 'react-native-router-flux'
import { cardType } from '../../Types/CardType'
import { getIndex } from '../../Realm/RealmService'
import Swiper from 'react-native-swiper'

class CardSwiper extends Component {
  renderCardDetails (card, key) {
    return (
      <CardDetails card={card} key={key} />
    )
  }

  updateScreenTitle (e, state) {
    const { index } = state
    const getCardName = (index) => this.children[index].props.card.name
    NavigationActions.refresh({title: `${getCardName(index)}`})
  }

  render () {
    const { cards, card } = this.props
    let { index } = this.props

    if (card !== null) {
      index = getIndex(cards, card)
    }

    return (
      <Swiper
        ref={(component) => { this.swiper = component }}
        onMomentumScrollEnd={this.updateScreenTitle}
        style={{ flex: 1 }}
        loadMinimal
        loadMinimalSize={2}
        loop={false}
        showsPagination={false}
        index={index}
        >
        {cards.map(this.renderCardDetails)}
      </Swiper>
    )
  }
}

CardSwiper.propTypes = {
  cards: PropTypes.objectOf(cardType),
  card: cardType,
  index: PropTypes.number
}

CardSwiper.defaultProps = {
  index: 0,
  card: null
}

export default CardSwiper
