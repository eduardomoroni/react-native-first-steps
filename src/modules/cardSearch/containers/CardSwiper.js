import React, { Component } from 'react'
import { View } from 'react-native'
import PropTypes from 'prop-types'
import CardDetails from './CardDetails'
import { Actions as NavigationActions } from 'react-native-router-flux'
import { cardType } from '../../../types/CardType'
import { findCardIndex } from '../../../services/CardService'
import Swiper from 'react-native-swiper'
import { FloatingActionButton } from '../../shared/components/index'

const fabItems = [
  { name: 'md-share',
    buttonColor: '#9b59b6',
    title: 'Share',
    onPress: () => null
  }, {
    name: 'md-folder-open',
    buttonColor: '#3498db',
    title: 'Add to Deck',
    onPress: () => null
  }, {
    name: 'md-swap',
    buttonColor: '#1abc9c',
    title: 'Add to TradeList',
    onPress: () => null
  }
]

export class CardSwiper extends Component {
  renderCardDetails (card, key) {
    return (
      <CardDetails card={card} key={key} />
    )
  }

  updateScreenTitle (e, state) {
    const { index } = state
    const getCardName = (index) => this.children[index].props.card.name
    NavigationActions.refresh({title: `${getCardName(index)}`})
  }

  render () {
    const { cards, card } = this.props
    let { index } = this.props

    if (card !== null) {
      index = findCardIndex(cards, card)
    }

    return (
      <View>
        <Swiper
          ref={(component) => { this.swiper = component }}
          onMomentumScrollEnd={this.updateScreenTitle}
          style={{ flex: 1 }}
          loadMinimal
          loadMinimalSize={2}
          loop={false}
          showsPagination={false}
          index={index}
          >
          {cards.map(this.renderCardDetails)}
        </Swiper>
        <FloatingActionButton items={fabItems} />
      </View>
    )
  }
}

CardSwiper.propTypes = {
  cards: PropTypes.objectOf(cardType),
  card: cardType,
  index: PropTypes.number
}

CardSwiper.defaultProps = {
  index: 0,
  card: null
}

export default CardSwiper
