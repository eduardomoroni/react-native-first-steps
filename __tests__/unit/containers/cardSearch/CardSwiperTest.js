import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import { CardSwiper } from '../../../../src/modules/cardSearch/containers/CardSwiper'
import Swiper from 'react-native-swiper'
import _ from 'lodash'
import RealmResult from '../../../assets/Stubs/RealmResult.json'
import * as RealmService from '../../../../src/services/realm'

const cardIndex = 1
const props = {
  cards: _.flatMap(RealmResult),
  card: RealmResult[cardIndex]
}

describe('<CardSwiper />', () => {
  RealmService.getIndex = jest.fn((cards, card) => cardIndex)

  const wrapper = shallow(<CardSwiper {...props} />)
  const cardSwiperInstance = wrapper.instance()
  const swiper = wrapper.find(Swiper)

  it('Should Render CardSwiper with card as initial slide', () => {
    expect(swiper.prop('index')).toBe(cardIndex)
    expect(cardSwiperInstance.props.card).toBe(props.card)
    expect(wrapper).toMatchSnapshot()
  })

  xit('Should change navBarTitle when swipe cards', () => {
    // Enzyme instance is behaving different in this component, FIGURE THIS OUT (LATER)
    expect(swiper.prop('onMomentumScrollEnd')).toBe(cardSwiperInstance.updateScreenTitle)
  })
})
