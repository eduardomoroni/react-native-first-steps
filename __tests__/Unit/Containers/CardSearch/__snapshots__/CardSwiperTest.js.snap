// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardSwiper /> Should Render CardSwiper with card as initial slide 1`] = `
<View>
  <_class
    automaticallyAdjustContentInsets={false}
    autoplay={false}
    autoplayDirection={true}
    autoplayTimeout={2.5}
    bounces={false}
    horizontal={true}
    index={1}
    loadMinimal={true}
    loadMinimalSize={2}
    loop={false}
    onMomentumScrollEnd={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={true}
    scrollsToTop={false}
    showsButtons={false}
    showsHorizontalScrollIndicator={false}
    showsPagination={false}
    showsVerticalScrollIndicator={false}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <CardDetails
      card={
        Object {
          "artist": "Zack Stella",
          "cmc": 2,
          "colorIdentity": Object {
            "0": Object {
              "colorIdentity": "U",
            },
          },
          "colors": Object {
            "0": Object {
              "color": "Blue",
            },
          },
          "flavor": "It is fascinating to see how the people of this plane replicate magical techniques without a natural gift.",
          "foreignNames": Object {
            "0": Object {
              "language": "Chinese Simplified",
              "multiverseid": 423891,
              "name": "失效",
            },
            "1": Object {
              "language": "Chinese Traditional",
              "multiverseid": 424075,
              "name": "失效",
            },
            "2": Object {
              "language": "French",
              "multiverseid": 424443,
              "name": "Négation",
            },
            "3": Object {
              "language": "German",
              "multiverseid": 424259,
              "name": "Negieren",
            },
            "4": Object {
              "language": "Italian",
              "multiverseid": 424627,
              "name": "Negare",
            },
            "5": Object {
              "language": "Japanese",
              "multiverseid": 424811,
              "name": "否認",
            },
            "6": Object {
              "language": "Korean",
              "multiverseid": 424995,
              "name": "부인",
            },
            "7": Object {
              "language": "Portuguese (Brazil)",
              "multiverseid": 425179,
              "name": "Negar",
            },
            "8": Object {
              "language": "Russian",
              "multiverseid": 425363,
              "name": "Отвергнуть",
            },
            "9": Object {
              "language": "Spanish",
              "multiverseid": 425547,
              "name": "Negar",
            },
          },
          "id": "66106dba089787a1d0d5fe1b80091e7eebe29e55",
          "imagename": "negate",
          "layout": "normal",
          "legalities": Object {
            "0": Object {
              "format": "Battle for Zendikar Block",
              "legality": "Legal",
            },
            "1": Object {
              "format": "Commander",
              "legality": "Legal",
            },
            "2": Object {
              "format": "Khans of Tarkir Block",
              "legality": "Legal",
            },
            "3": Object {
              "format": "Legacy",
              "legality": "Legal",
            },
            "4": Object {
              "format": "Lorwyn-Shadowmoor Block",
              "legality": "Legal",
            },
            "5": Object {
              "format": "Modern",
              "legality": "Legal",
            },
            "6": Object {
              "format": "Standard",
              "legality": "Legal",
            },
            "7": Object {
              "format": "Vintage",
              "legality": "Legal",
            },
          },
          "manaCost": "{1}{U}",
          "multiverseid": 423707,
          "name": "Negate",
          "number": "40",
          "originalText": "Counter target noncreature spell.",
          "originalType": "Instant",
          "power": null,
          "printings": Object {
            "0": Object {
              "printing": "pMPR",
            },
            "1": Object {
              "printing": "MOR",
            },
            "10": Object {
              "printing": "ORI",
            },
            "11": Object {
              "printing": "OGW",
            },
            "12": Object {
              "printing": "CN2",
            },
            "13": Object {
              "printing": "AER",
            },
            "2": Object {
              "printing": "M10",
            },
            "3": Object {
              "printing": "DPA",
            },
            "4": Object {
              "printing": "M11",
            },
            "5": Object {
              "printing": "M12",
            },
            "6": Object {
              "printing": "M13",
            },
            "7": Object {
              "printing": "M14",
            },
            "8": Object {
              "printing": "M15",
            },
            "9": Object {
              "printing": "DTK",
            },
          },
          "rarity": "Common",
          "rulings": Object {
            "0": Object {
              "date": "2009-10-01",
              "text": "A creature spell is any spell with the type creature, even if it has other types such as artifact or enchantment. Older cards of type summon are also creature spells.",
            },
            "1": Object {
              "date": "2009-10-01",
              "text": "Testing.",
            },
          },
          "subtypes": Object {},
          "superTypes": Object {},
          "text": "Counter target noncreature spell.",
          "toughness": null,
          "type": "Instant",
          "types": Object {
            "0": Object {
              "type": "Instant",
            },
          },
        }
      }
    />
    <CardDetails
      card={
        Object {
          "artist": "Jakub Kasper",
          "cmc": 2,
          "colorIdentity": Object {
            "0": Object {
              "colorIdentity": "W",
            },
          },
          "colors": Object {
            "0": Object {
              "color": "White",
            },
          },
          "flavor": "This little critter can turn all of you to scrap in under thirty seconds! Come on, who wants to be the first to go? Try and stop me!",
          "foreignNames": Object {
            "0": Object {
              "language": "Chinese Simplified",
              "multiverseid": 423858,
              "name": "大胆渗透者",
            },
            "1": Object {
              "language": "Chinese Traditional",
              "multiverseid": 424042,
              "name": "大膽滲透者",
            },
            "2": Object {
              "language": "French",
              "multiverseid": 424410,
              "name": "Infiltrateur audacieux",
            },
            "3": Object {
              "language": "German",
              "multiverseid": 424226,
              "name": "Dreister Eindringling",
            },
            "4": Object {
              "language": "Italian",
              "multiverseid": 424594,
              "name": "Infiltrato Audace",
            },
            "5": Object {
              "language": "Japanese",
              "multiverseid": 424778,
              "name": "大胆な潜入者",
            },
            "6": Object {
              "language": "Korean",
              "multiverseid": 424962,
              "name": "무모한 침투자",
            },
            "7": Object {
              "language": "Portuguese (Brazil)",
              "multiverseid": 425146,
              "name": "Infiltrador Audacioso",
            },
            "8": Object {
              "language": "Russian",
              "multiverseid": 425330,
              "name": "Дерзкий Лазутчик",
            },
            "9": Object {
              "language": "Spanish",
              "multiverseid": 425514,
              "name": "Infiltrador audaz",
            },
          },
          "id": "3886f500667c6506f8b835843bc39a10a0d8f2ec",
          "imagename": "audacious infiltrator",
          "layout": "normal",
          "legalities": Object {},
          "manaCost": "{1}{W}",
          "multiverseid": 423674,
          "name": "Audacious Infiltrator",
          "number": "7",
          "originalText": "Audacious Infiltrator can\\"t be blocked by artifact creatures.",
          "originalType": "Creature — Dwarf Rogue",
          "power": 3,
          "printings": Object {
            "0": Object {
              "printing": "AER",
            },
          },
          "rarity": "Common",
          "rulings": Object {},
          "subtypes": Object {
            "0": Object {
              "subType": "Dwarf",
            },
            "1": Object {
              "subType": "Rogue",
            },
          },
          "superTypes": Object {},
          "text": "Audacious Infiltrator can\\"t be blocked by artifact creatures.",
          "toughness": 1,
          "type": "Creature — Dwarf Rogue",
          "types": Object {
            "0": Object {
              "type": "Creature",
            },
          },
        }
      }
    />
  </_class>
  <FloatingActionButton
    items={
      Array [
        Object {
          "buttonColor": "#9b59b6",
          "name": "md-share",
          "onPress": [Function],
          "title": "Share",
        },
        Object {
          "buttonColor": "#3498db",
          "name": "md-folder-open",
          "onPress": [Function],
          "title": "Add to Deck",
        },
        Object {
          "buttonColor": "#1abc9c",
          "name": "md-swap",
          "onPress": [Function],
          "title": "Add to TradeList",
        },
      ]
    }
  />
</View>
`;
